lib/vanagon/optparse.rb

```diff
@@ -2,40 +2,87 @@
 
 class Vanagon
   class OptParse
-    FLAGS = {
-        :workdir => ['-w DIR', '--workdir DIR', "Working directory where build source should be put (defaults to a tmpdir)"],
-        :"remote-workdir" => ['-r DIR', '--remote-workdir DIR', "Working directory where build source should be put on the remote host (defaults to a tmpdir)"],
-        :configdir => ['-c', '--configdir DIR', 'Configs dir (defaults to $pwd/configs)'],
-        :target => ['-t HOST', '--target HOST', 'Configure a target machine for build and packaging (defaults to grabbing one from the pooler)'],
-        :engine => ['-e ENGINE', '--engine ENGINE', "A custom engine to use (defaults to the pooler) [base, local, docker, pooler currently supported]"],
-        :skipcheck => ['--skipcheck', 'Skip the `check` stage when building components'],
-        :preserve => ['-p', '--preserve', 'Whether to tear down the VM on success or not (defaults to false)'],
-        :verbose => ['-v', '--verbose', 'Verbose output (does nothing)'],
-        :only_build => ["--only-build COMPONENT,COMPONENT,...", Array, 'Only build this array of components']
-      }.freeze
-
-    def initialize(banner, options = [])
-      @options = Hash[options.map { |v| [v, nil] }]
-      @optparse = OptionParser.new do |opts|
+    def initialize(banner, symbols = []) # rubocop:disable Metrics/AbcSize
+      ## EXG -- don't understand why symbols array is needed.
```

Comment:
oh no I think I found what the symbols array is doing ... it looks like it's setting the valid options for each of the subcommands and only adding the option if it's listed as one of the supported symbols

Reply:
Ah ha. That explains the confusing original implementation.

I'm tempted then to monkey-patch #on to take the symbols array as an argument and handle it that way.

Reply:
I've so far avoided monkey patches in ruby (someone scared me early on in my ruby days), are there non-terrible ways to do this?

Reply:
IMO, there's a balance here between simplicity (allowing the parser to parser whatever, even if it doesn't apply for the subcommand) and the cost of accuracy.

The classic, non-monkey-patch, approach is to wrap or subclass the OptionsParser object. My preferences, in descending order.

1. Opt for simplicity. Let the subcommand ignore stuff not relevant to it.
2. Monkey patch #on. This, IMO, is the textbook use-case for it.
3. Wrap OptionsParser. Major advantage is that this pattern is very easy to recognize and contains the least surprise over monkey patching.
4. Subclass OptionsParser. Uggh.

I'm happy to consider alternate opinions.

Reply:
Cool, thanks for the explanation! I think option 1 makes a lot of sense, it doesn't seem particularly dangerous to be ignoring some of the options, and a lot of that can be clarified in the README in the future.

Reply:
Of course a further thought -- replacing everything with docopt could be valuable.

```ruby
options[subcommand] = <<~ENDDOC
  <options description for subcommand>
ENDDOC ...

options = Docopt::docopt(options[subcommand])
```

However, this would case a necessary change in all the consumers. For the sake of 1 switch this feels like an even bigger rabbit hole than I've found myself in.

Reply:
As of now, the 'opt-for-simplicity' is the leading option.